indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java

```diff
@@ -0,0 +1,685 @@
+    {
+      log.info("Adding request to batch [%s]: %s", key, request.getAction());
+      return requestToFuture.computeIfAbsent(request, req -> new CompletableFuture<>());
+    }
+
+    synchronized void merge(AllocateRequestBatch batch)
+    {
+      requestToFuture.putAll(batch.requestToFuture);
+      batch.requestToFuture.clear();
+    }
+
+    synchronized List<SegmentAllocateRequest> getRequests()
+    {
+      return new ArrayList<>(requestToFuture.keySet());
+    }
+
+    synchronized void markCompleted()
+    {
+      if (!requestToFuture.isEmpty()) {
+        log.info("Marking [%d] requests in batch [%s] as failed.", size(), key);
+        requestToFuture.values().forEach(future -> future.complete(null));
+        requestToFuture.keySet().forEach(
+            request -> emitTaskMetric("task/action/failed/count", 1L, request)
+        );
+        requestToFuture.clear();
+      }
+    }
+
+    synchronized void handleResult(SegmentAllocateResult result, SegmentAllocateRequest request)
+    {
+      request.incrementAttempts();
+
+      if (result.isSuccess()) {
+        emitTaskMetric("task/action/success/count", 1L, request);
+        requestToFuture.remove(request).complete(result.getSegmentId());
+        return;
+      }
+
+      log.info("Failed to allocate segment for action [%s]: %s", request.getAction(), result.getErrorMessage());
+      if (request.canRetry()) {
+        log.debug(
+            "Can requeue action [%s] after [%d] failed attempts.",
+            request.getAction(),
+            request.getAttempts()
+        );
+      } else {
+        emitTaskMetric("task/action/failed/count", 1L, request);
+        log.error(
+            "Removing allocation action [%s] from batch after [%d] failed attempts.",
+            request.getAction(),
+            request.getAttempts()
+        );
+        requestToFuture.remove(request).complete(null);
+      }
+    }
+
+    synchronized boolean isEmpty()
+    {
+      return requestToFuture.isEmpty();
+    }
+
+    synchronized int size()
+    {
+      return requestToFuture.size();
+    }
+
+    void resetQueueTime()
+    {
+      queueTimeMillis = System.currentTimeMillis();
+    }
+
+    long getQueueTime()
+    {
+      return queueTimeMillis;
+    }
+
+    boolean isDue()
+    {
+      return System.currentTimeMillis() - queueTimeMillis > maxWaitTimeMillis;
+    }
+  }
+
+  /**
+   * Key to identify a batch of allocation requests.
+   */
+  private static class AllocateRequestKey
+  {
+    private final boolean unique;
+
+    private final String dataSource;
+    private final String groupId;
+    private final Interval preferredAllocationInterval;
+    private final Granularity preferredSegmentGranularity;
+
+    private final boolean skipSegmentLineageCheck;
+    private final LockGranularity lockGranularity;
+
+    private final boolean useNonRootGenPartitionSpace;
+
+    private final int hash;
+    private final String serialized;
```

Comment:
does this string need to be stored? Have you thought about possible GC impact of storing these requests as batches? what will happen if segment allocation is slow for some reason? would Overlord continue to take in requests?

Reply:
> does this string need to be stored?

No, I think we can get rid of this. Not sure why I had done this in the first place. The computations in a batch are heavy enough to not make this matter at all. 😅

Reply:
> Have you thought about possible GC impact of storing these requests as batches?

Thanks for the reminder, I do need to do some basic GC evaluation.

Reply:
> what will happen if segment allocation is slow for some reason? would Overlord continue to take in requests?

Currently, yes. We can try to put a limit on the queue size and/or total number of requests in queue. What do you think?

Reply:
yeah. we don't want the Overlord to crash with an OOM. So having some guardrail will be better.

Reply:
Put a limit of 5000 on the queue size.